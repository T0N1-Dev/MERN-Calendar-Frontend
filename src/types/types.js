export const types = {

    uiOpenModal: '[ui] Open modal',
    uiCloseModal: '[ui] Close modal',
    uiSetError: '[ui] Set error',

    eventSetActive: '[event] Set active',
    eventStartAddNew: '[event] Start add new',
    eventAddNew: '[event] Add new',
    eventUpdate: '[event] Event updated',
    eventDelete: '[event] Event deleted',
    eventLoaded: '[event] Events loaded',

    authCheckingFinish: '[auth] Finish checking login state',
    authStartLogin: '[auth] Start login',
    authLogin: '[auth] Login',
    authStartRegister: '[auth] Start Register',
    authStartTokenRenew: '[auth] Start token renew',
    authLogout: '[auth] Logout',
}