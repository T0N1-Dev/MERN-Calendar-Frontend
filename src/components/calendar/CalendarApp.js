import React, { useEffect, useState } from 'react'
import { formatDate } from '@fullcalendar/core'
import FullCalendar from '@fullcalendar/react'
import dayGridPlugin from '@fullcalendar/daygrid'
import timeGridPlugin from '@fullcalendar/timegrid'
import interactionPlugin from '@fullcalendar/interaction'
import { createEventId } from './event-utils'
import Swal from 'sweetalert2'
import './calendar.css'
import { RoundButton } from '../ui/RoundButton'
import { eventStartAddNewEvent, eventStartDelete, eventStartLoading, eventStartUpdate } from "../../actions/events";
import { useDispatch, useSelector } from 'react-redux'

export default function CalendarApp() {

  const [weekendsVisible, setWeekendsVisible] = useState(true)
  const [currentEvents, setCurrentEvents] = useState([])
  const dispatch = useDispatch();
  const { events } = useSelector((state) => state.calendar);
  const { color } = useSelector((state) => state.auth);

  useEffect(() => {
    dispatch( eventStartLoading() );
  }, [dispatch]);

  function handleWeekendsToggle() {
    setWeekendsVisible(!weekendsVisible)
  }

  function handleDateSelect(selectInfo) {

    const { start, end, allDay } = selectInfo;
    const backgroundColor = color;

    Swal.fire({
      title: "Please enter a new title for your event",
      input: "text",
      inputAttributes: {
        autocapitalize: "on"
      },
      showCancelButton: true,
      confirmButtonText: "Insert",
      showLoaderOnConfirm: true,
      preConfirm: (title) => {
        const event = { title, start, end, allDay, backgroundColor };
        console.log("Default Event", event);
        dispatch(eventStartAddNewEvent(event))
      },
      allowOutsideClick: () => !Swal.isLoading()
    }).then((result) => {
      if (result.isConfirmed) {
        let title = result.value.login;
        let calendarApi = selectInfo.view.calendar;

        calendarApi.unselect() // clear date selection

        if (title) {
          calendarApi.addEvent({
            id: createEventId(),
            title,
            start: selectInfo.startStr,
            end: selectInfo.endStr,
            allDay: selectInfo.allDay
          })
        }
      }
    });
  }

  function handleEventClick(clickInfo) {
    Swal.fire({
      title: 'Are you sure?',
      text: `Do you want to delete the event '${clickInfo.event.title}'?`,
      icon: 'warning',
      showCancelButton: true,
      confirmButtonText: 'Yes, delete it!',
      cancelButtonText: 'Cancel'
    }).then((result) => {
      if (result.isConfirmed) {
        dispatch(eventStartDelete(clickInfo.event.id));
        clickInfo.event.remove()
        Swal.fire({
          title: "Correct!",
          icon: "success",
          draggable: true
        });
      }
    })
  }

  function handleEvents(events) {
    setCurrentEvents(events)
  }

  function handleUpdate(update){
    const { id, title, start, end, allDay, backgroundColor } = update.event;
    const eventUpdated = {id, title, start, end, allDay, backgroundColor};
    dispatch(eventStartUpdate(eventUpdated));
  }

  async function handleRoundedButtonClick() {
    const { value: formValues } = await Swal.fire({
      title: "Add new event",
      html: `
        <input id="title" class="swal2-input" placeholder="Enter event title" required>
  
        <label for="start-date" style="display:block; text-align:center; margin-top:20px;">Start</label>
        <input id="start-date" class="swal2-input" type="date" style="margin-top:0;">
        <label for="end-date" style="display:block; text-align:center; margin-top:20px;">End</label>
        <input id="end-date" class="swal2-input" type="date" style="margin-top:0;">
      `,
      focusConfirm: false,
      preConfirm: () => {
        const titleInput = document.getElementById("title");
        const startDate = document.getElementById("start-date");
        const endDate = document.getElementById("end-date").value;

        if (!titleInput.value.trim()) {
          titleInput.style.border = "2px solid red";
          return false;
        } else if (!startDate.value) {
          startDate.style.border = "2px solid red";
          return false;
        }

        return { 
          titleInput,
          startDate,
          endDate
        };
      }
    });

    const { titleInput, startDate, endDate } = formValues || {};

    if (titleInput && startDate) {
      const title = titleInput.value;
      const start = new Date(`${startDate.value}T00:00:00`);
      const end = endDate ? new Date(`${endDate}T00:00:00`) : null;
      const newEvent = { title, start, end, allDay:true, backgroundColor: color };
      dispatch(eventStartAddNewEvent(newEvent));

      Swal.fire({
        title: "Great!",
        icon: "success",
        draggable: true
      });
    } 
  }
  
  const handleEventHover = (info) => {
      const username = info.event.extendedProps.user.username;
  
      const tooltip = document.createElement('div');
      tooltip.innerText = `By ${username}`;
      tooltip.style.position = 'absolute';
      tooltip.style.height = '25px';
      tooltip.style.backgroundColor = '#938b8d';
      tooltip.style.color = 'white';
      tooltip.style.padding = '4px 8px';
      tooltip.style.borderRadius = '4px';
      tooltip.style.fontSize = '12px';
      tooltip.style.zIndex = '1000';
  
      const { left, top, height } = info.el.getBoundingClientRect();
      tooltip.style.left = `${left}px`;
      tooltip.style.top = `${top + height + window.scrollY + 4}px`;
  
      document.body.appendChild(tooltip);
  
      info.el.addEventListener('mouseleave', () => {
        tooltip.remove();
      }, { once: true });
  }
  

  return (
    <div className='demo-app'>
      <Sidebar
        weekendsVisible={weekendsVisible}
        handleWeekendsToggle={handleWeekendsToggle}
        currentEvents={currentEvents}
      />
      <div className='demo-app-main'>
        <FullCalendar
          plugins={[dayGridPlugin, timeGridPlugin, interactionPlugin]}
          headerToolbar={{
            left: 'prev,next today',
            center: 'title',
            right: 'dayGridMonth,timeGridWeek,timeGridDay'
          }}
          validRange={{
            start: '2024-01-01',
            end: '2027-01-01'
          }}
          eventMouseEnter={handleEventHover}
          initialView='dayGridMonth'
          editable={true}
          selectable={true}
          selectMirror={true}
          dayMaxEvents={true}
          weekends={weekendsVisible}
          events={events}
          select={handleDateSelect}
          eventContent={renderEventContent}
          eventClick={handleEventClick}
          eventsSet={handleEvents}
          eventChange={handleUpdate}
          
        />
        <RoundButton onClick={handleRoundedButtonClick} />
      </div>
    </div>
  )
}

function renderEventContent(eventInfo) {
  return (
    <>
      <b>{eventInfo.timeText}</b>
      <i>{eventInfo.event.title}</i>
    </>
  )
}

function Sidebar({ weekendsVisible, handleWeekendsToggle, currentEvents }) {
  return (
    <div className='demo-app-sidebar'>
      <div className='demo-app-sidebar-section'>
        <h2>Instructions</h2>
        <ul>
          <li>Select dates and you will be ready to create a new event</li>
          <li>Drag, drop, and resize events</li>
          <li>Click an event to delete it</li>
        </ul>
      </div>
      <div className='demo-app-sidebar-section'>
        <label>
          <input
            type='checkbox'
            checked={weekendsVisible}
            onChange={handleWeekendsToggle}
          ></input>
          toggle weekends
        </label>
      </div>
      <div className='demo-app-sidebar-section'>
        <h2>All Events ({currentEvents.length})</h2>
        <ul>
          {currentEvents.map((event) => (
            <SidebarEvent key={event.id} event={event} />
          ))}
        </ul>
      </div>
    </div>
  )
}

function SidebarEvent({ event }) {
  return (
    <li key={event.id}>
      <b>{formatDate(event.start, {year: 'numeric', month: 'short', day: 'numeric'})}</b>
      <i>{event.title}</i>
    </li>
  )
}