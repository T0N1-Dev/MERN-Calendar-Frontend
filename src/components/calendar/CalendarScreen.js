import React from 'react';
import { Navbar } from '../ui/Navbar';
import "@mobiscroll/react/dist/css/mobiscroll.min.css";
import {
  Button,
  Datepicker,
  Dropdown,
  Eventcalendar,
  Input,
  Popup,
  Segmented,
  SegmentedGroup,
  setOptions,
  Snackbar,
  Switch,
  Textarea,
} from '@mobiscroll/react';
import { useCallback, useMemo, useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { uiOpenModal, uiCloseModal } from '../../actions/ui';
import { eventStartAddNewEvent, eventStartUpdate, eventStartDelete, eventStartLoading, eventLoaded, eventStartSetActive } from '../../actions/events';
import { RoundButton } from '../ui/RoundButton';
import _ from 'lodash';

setOptions({
  theme: 'windows',
  themeVariant: 'light'
});

export const CalendarScreen = () => {
  const { events, activeEvent } = useSelector((state) => state.calendar);
  const { modalOpen, mgError } = useSelector((state) => state.ui);
  const { uid, name, color } = useSelector((state) => state.auth);
  const [tempEvent, setTempEvent] = useState(null);
  const [undoEvent, setUndoEvent] = useState(null);
  const [isEdit, setEdit] = useState(false);
  const [anchor, setAnchor] = useState(null);
  const [start, startRef] = useState(null);
  const [end, endRef] = useState(null);
  const [popupEventTitle, setTitle] = useState('');
  const [popupEventDescription, setDescription] = useState('');
  const [popupEventAllDay, setAllDay] = useState(true);
  const [popupTravelTime, setTravelTime] = useState(0);
  const [popupEventDate, setDate] = useState([]);
  const [popupEventStatus, setStatus] = useState('busy');
  const [isSnackbarOpen, setSnackbarOpen] = useState(false);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch( eventStartLoading() );
    if (mgError) {
      setSnackbarOpen(false);
    }
  }, [dispatch, mgError]);

  // It's necessary to avoid a mutation of the Redux state by the EventCalendar component
  const eventCalendarData = useMemo(() => _.cloneDeep(events), [events]);

  const snackbarButton = useMemo(
    () => ({
      action: () => {
        dispatch(eventStartAddNewEvent(undoEvent));
      },
      text: 'Undo',
    }),
    [undoEvent, dispatch],
  );

  const handleSnackbarClose = useCallback(() => {
    setSnackbarOpen(false);
  }, []);

  const saveEvent = useCallback(() => {
    const newEvent = {
      id: '',
      title: popupEventTitle,
      description: popupEventDescription,
      start: popupEventDate[0],
      end: popupEventDate[1],
      allDay: popupEventAllDay,
      bufferBefore: popupTravelTime,
      status: popupEventStatus,
      color: color,
      tooltip: 'By ' + name,
      user: {uid, name}
    };
    if (isEdit) {
      newEvent.id = activeEvent.id;
      dispatch(eventStartUpdate(newEvent));
    } else {
      dispatch(eventStartAddNewEvent(newEvent));
    }
    // close the popup
    dispatch(uiCloseModal());
  }, [
    isEdit,
    popupEventAllDay,
    popupEventDate,
    popupEventDescription,
    popupEventStatus,
    popupEventTitle,
    popupTravelTime,
    uid,
    name,
    color,
    activeEvent,
    dispatch
  ]);

  const deleteEvent = useCallback(
    (event) => {
      dispatch(eventStartDelete());
      if (!mgError) {
        setUndoEvent(event);
        setSnackbarOpen(true);
      }
    },
    [dispatch, mgError],
  );

  const loadPopupForm = useCallback((event) => {
    setTitle(event.title);
    setDescription(event.description);
    setDate([event.start, event.end]);
    setAllDay(event.allDay || false);
    setTravelTime(event.bufferBefore || 0);
    setStatus(event.status || 'busy');
  }, []);

  const titleChange = useCallback((ev) => {
    setTitle(ev.target.value);
  }, []);

  const descriptionChange = useCallback((ev) => {
    setDescription(ev.target.value);
  }, []);

  const allDayChange = useCallback((ev) => {
    setAllDay(ev.target.checked);
  }, []);

  const travelTimeChange = useCallback((ev) => {
    setTravelTime(ev.target.value);
  }, []);

  const dateChange = useCallback((args) => {
    setDate(args.value);
  }, []);

  const statusChange = useCallback((ev) => {
    setStatus(ev.target.value);
  }, []);

  const onDeleteClick = useCallback(() => {
    deleteEvent(tempEvent);
    dispatch(uiCloseModal());
  }, [deleteEvent, tempEvent, dispatch]);

  const onEventClick = useCallback(
    (args) => {
      setEdit(true);
      !_.isEqual(args.event, activeEvent) && dispatch(eventStartSetActive({...args.event}));
      setTempEvent(args.event);
      // fill popup form with event data
      loadPopupForm(args.event);
      setAnchor(args.domEvent.target);
      dispatch(uiOpenModal());
    },
    [loadPopupForm, dispatch, activeEvent],
  );

  const onEventCreated = useCallback(
    (args) => {
      setEdit(false);
      setTempEvent(args.event);
      // fill popup form with event data
      loadPopupForm(args.event);
      setAnchor(args.target);
      // open the popup
      dispatch(uiOpenModal());
    },
    [loadPopupForm, dispatch],
  );

  const onEventDeleted = useCallback(
    (args) => {
      deleteEvent(args.event);
    },
    [deleteEvent],
  );

  const onEventUpdated = useCallback((args) => {
    dispatch(eventStartUpdate({...args.event}));
  }, [dispatch]);

  const handleCancelClick = useCallback(() => {
    if ( tempEvent?.id !== activeEvent?.id ) {
      dispatch(eventLoaded(events.filter(event => event.id !== tempEvent?.id)));
    }
    dispatch(uiCloseModal());
  }, [dispatch, events, tempEvent, activeEvent]);

  const controls = useMemo(() => (popupEventAllDay ? ['date'] : ['datetime']), [popupEventAllDay]);
  const datepickerResponsive = useMemo(
    () =>
      popupEventAllDay
        ? {
            medium: {
              controls: ['calendar'],
              touchUi: false,
            },
          }
        : {
            medium: {
              controls: ['calendar', 'time'],
              touchUi: false,
            },
          },
    [popupEventAllDay],
  );

  const headerText = useMemo(() => (isEdit ? 'Edit event' : 'New Event'), [isEdit]);
  const popupButtons = useMemo(() => {
    if (isEdit) {
      return [
        {
          handler: handleCancelClick,
          keyCode: 'escape',
          text: 'Cancel',
          cssClass: 'mbsc-popup-button-secondary'
        },
        {
          handler: () => {
            saveEvent();
          },
          keyCode: 'enter',
          text: 'Save',
          cssClass: 'mbsc-popup-button-primary',
        },
      ];
    } else {
      return [
        {
          handler: handleCancelClick,
          keyCode: 'escape',
          text: 'Cancel',
          cssClass: 'mbsc-popup-button-secondary'
        },
        {
          handler: () => {
            saveEvent();
          },
          keyCode: 'enter',
          text: 'Add',
          cssClass: 'mbsc-popup-button-primary',
        },
      ];
    }
  }, [isEdit, saveEvent, handleCancelClick]);

  const popupResponsive = useMemo(
    () => ({
      medium: {
        display: 'anchored',
        width: 400,
        fullScreen: false,
        touchUi: false,
      },
    }),
    [],
  );

  const onClose = useCallback(() => {
    if ( tempEvent?.id !== activeEvent?.id ) {
      dispatch(eventLoaded(events.filter(event => event.id !== tempEvent?.id)));
    }
    dispatch(uiCloseModal());
  }, [dispatch, events, tempEvent, activeEvent ]);

  const handleRoundButtonClick = useCallback(() => {
    const newEvent = {
      id: '',
      title: 'New Event',
      notes: '',
      start: new Date(),
      end: new Date(),
      allDay: true,
      bufferBefore: 0,
      status: 'busy',
      color: color,
      tooltip: `By ${name}`,
      user: { uid, name }
    };
    dispatch(eventStartSetActive({...newEvent}));
    setAnchor(null);
    setEdit(false);
    setTempEvent(newEvent);
    loadPopupForm(newEvent);
    setAnchor(null);
    dispatch(uiOpenModal());
  }, [color, uid, name, loadPopupForm, dispatch]);

  console.log(eventCalendarData);
  return (
    <>
      <Navbar />
      <Eventcalendar
        clickToCreate={true}
        // immutableData={true} // good idea to avoid eventCalendarData (a copy of events)
        dragToCreate={true}
        dragToMove={true}
        dragToResize={true}
        data={eventCalendarData}
        onEventClick={onEventClick}
        onEventCreated={onEventCreated}
        onEventDeleted={onEventDeleted}
        onEventUpdated={onEventUpdated}
        data-testid="event-calendar-mock"
      />
      <Popup
        display="center"
        contentPadding={false}
        headerText={headerText}
        anchor={anchor}
        buttons={popupButtons}
        isOpen={modalOpen}
        onClose={onClose}
        responsive={popupResponsive}
        touchUi={false}
      >
        <div className="mbsc-form-group">
          <Input label="Title" value={popupEventTitle} onChange={titleChange} />
          <Textarea label="Description" value={popupEventDescription} onChange={descriptionChange} />
        </div>
        <div className="mbsc-form-group">
          <Switch label="All-day" checked={popupEventAllDay} onChange={allDayChange} />
          <Input ref={startRef} label="Starts" />
          <Input ref={endRef} label="Ends" />
          {!popupEventAllDay && (
            <div id="travel-time-group">
              <Dropdown label="Travel time" value={popupTravelTime} onChange={travelTimeChange}>
                <option value="0">None</option>
                <option value="5">5 minutes</option>
                <option value="15">15 minutes</option>
                <option value="30">30 minutes</option>
                <option value="60">1 hour</option>
                <option value="90">1.5 hours</option>
                <option value="120">2 hours</option>
              </Dropdown>
            </div>
          )}
          <Datepicker
            select="range"
            display="anchored"
            controls={controls}
            touchUi={true}
            startInput={start}
            endInput={end}
            showRangeLabels={false}
            responsive={datepickerResponsive}
            onChange={dateChange}
            value={popupEventDate}
          />
          <SegmentedGroup onChange={statusChange}>
            <Segmented value="busy" checked={popupEventStatus === 'busy'}>
              Show as busy
            </Segmented>
            <Segmented value="free" checked={popupEventStatus === 'free'}>
              Show as free
            </Segmented>
          </SegmentedGroup>
          {isEdit ? (
            <div className="mbsc-button-group">
              <Button className="mbsc-button-block" color="danger" variant="outline" onClick={onDeleteClick} data-testid="delete-event">
                Delete event
              </Button>
            </div>
          ) : null}
        </div>
      </Popup>
      <Snackbar isOpen={isSnackbarOpen} message="Event deleted" button={snackbarButton} onClose={handleSnackbarClose} />
      <RoundButton onClick={handleRoundButtonClick}/>
    </>
  );
}