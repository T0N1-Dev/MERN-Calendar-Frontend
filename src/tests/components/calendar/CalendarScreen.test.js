import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { CalendarScreen } from "../../../components/calendar/CalendarScreen";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { calendarReducer } from "../../../reducers/calendarReducer";
import { authReducer } from "../../../reducers/authReducer";
import { uiReducer } from "../../../reducers/uiReducer";
import { types } from "../../../types/types";

jest.mock("@mobiscroll/react", () => ({
    Eventcalendar: ({ data }) => (
        <div data-testid="event-calendar-mock">
            {data && data.length > 0 ? (
                data.map((event) => (
                    <div key={event.id} className="mock-event">
                        {event.title}
                    </div>
                ))
            ) : (
                <p>No events</p>
            )}
        </div>
    ),
    setOptions: jest.fn(),
    Button: ({ onClick, children }) => <button onClick={onClick}>{children}</button>,
    Input: () => <input />,
    Popup: () => <div>Mocked Popup</div>,
    Datepicker: () => <div>Mocked Datepicker</div>,
    Switch: () => <div>Mocked Switch</div>,
    Snackbar: () => <div>Mocked Snackbar</div>,
    Textarea: () => <textarea />,
    Dropdown: () => <div>Mocked Dropdown</div>,
    Segmented: () => <div>Mocked Segmented</div>,
    SegmentedGroup: () => <div>Mocked SegmentedGroup</div>,
}));  

describe('Testing on <CalendarScreen />', () => {
    
    let store;
    let testContainer;

    beforeEach(() => {
        HTMLCanvasElement.prototype.getContext = jest.fn();
        store = configureStore({
            reducer: {
                calendar: calendarReducer,
                ui: uiReducer,
                auth: authReducer
            },
            preloadedState: {
                calendar: { 
                    events: [{
                        title: 'Test Event',
                        start: new Date(),
                        end: new Date(),
                        allDay: true,
                        bufferBefore: 0,
                        status: 'busy',
                        color: 'blue',
                        tooltip: 'By Test',
                        user: null,
                        id: '123'
                  }], 
                  activeEvent: null 
                },
                ui: { modalOpen: false, mgError: null },
                auth: { uid: "123", name: "Test User", color: "blue" }
            }
        })

        store.dispatch = jest.fn();

        const {container} = render(
            <Provider store={store}>
                <CalendarScreen />
            </Provider>
        );

        testContainer = container;
    });
    
    test('should match Snapshot', () => {
        expect(testContainer).toMatchSnapshot();
    });   
    
    test("should show event after loading", async () => {
        const eventElement = await screen.findByTestId('event-calendar-mock');
        console.log(eventElement);
    });
    
});