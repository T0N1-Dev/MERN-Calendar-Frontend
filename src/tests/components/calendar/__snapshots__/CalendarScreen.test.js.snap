// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing on <CalendarScreen /> should match Snapshot 1`] = `
<div>
  <div
    class="navbar navbar-dark bg-dark mb-4"
  >
    <span
      class="navbar-brand mx-3"
    >
      <i
        class="fa-solid fa-user"
      />
        
      Test User
    </span>
    <div
      class="ml-auto"
    >
      <button
        class="btn btn-dark mx-3"
      >
        <i
          class="fa-solid fa-right-from-bracket"
        />
          
        <span>
          Logout
        </span>
      </button>
    </div>
  </div>
  <div
    data-testid="event-calendar-mock"
  >
    <div
      class="mock-event"
    >
      Test Event
    </div>
  </div>
  <div>
    Mocked Popup
  </div>
  <div>
    Mocked Snackbar
  </div>
  <button
    class="round-button"
  >
    <div
      class="pulse"
    />
    <i
      class="fa-solid fa-plus"
    />
  </button>
</div>
`;
